### PR TITLE
Periodic timers ⏰

### PR DESCRIPTION
**Pull request by alecdotninja**
This PR got a little out of control (:see_no_evil:), but it can be reviewed by commit.

It adds a new periodic timer API to components:
```ruby
class CountdownComponent < ViewComponent::Base
  include Motion::Component

  attr_reader :count

  def initialize(count: 10)
    @count = count
  end

  def call
    content_tag(:div) { "The count is #{count}." }
  end

  # This is sugar for `periodic_timer :tick, every: 1.second`.
  every 1.second, :tick

  def tick
    @count -= 1

    stop_periodic_timer :tick if count.zero?
  end
end
```

(Also, the testing here is minimal, but I intent to build it up more over time)

### CODE REVIEW

File: lib/motion/component/motions.rb [hunk 1]
```diff
@@ -10,6 +11,23 @@ module Component
     module Motions
       extend ActiveSupport::Concern
 
+      # Analogous to `module_function` (available on both class and instance)
+      module ModuleFunctions
```
Comment (Gmfholley) [inline lib/motion/component/motions.rb:15]: Nice!

File: lib/motion/action_cable_extentions/declarative_notifications.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+require "motion"
+
+module Motion
+  module ActionCableExtentions
+    # Provides a `periodicly_notify(broadcasts, to:)` API that can be used to
```
Comment (Gmfholley) [inline lib/motion/action_cable_extentions/declarative_notifications.rb:7]: Small typo: periodically

File: lib/motion/channel.rb [hunk 3]
```diff
@@ -56,10 +57,19 @@ def process_broadcast(broadcast, message)
...
+    end
+
     private
 
     def synchronize
-      streaming_from(component_connection.broadcasts, to: :process_broadcast)
+      streaming_from component_connection.broadcasts,
+        to: :process_broadcast
+
+      periodicly_notify component_connection.periodic_timers,
```
Comment (Gmfholley) [inline lib/motion/channel.rb:71]: typo: periodically

**Comment by alecdotninja** [comment]
Thanks for the feedback, Wendy! :heart: 

You're right; documentation is seriously lacking right now, and it's something that I want to improve. In my mind, I am waiting for the "Write the docs" story to fully document, but it is probably a good idea to add a blurb to the README in the meantime.

> Does adding this feature make it harder to swap out ActionCable for AnyCable ?
Yes, right now, timers are [not supported in AnyCable](https://docs.anycable.io/v1/#/ruby/compatibility) (which makes sense given its design).

It is worth noting though that right now we have _no_ support for AnyCable. I should have PR with basic support (no timers or broadcasts) out soon.

> If we don't implement this, does this mean the user can't come up with solutions to this kind of problem, or does this just mean they have to implement it themselves?

The only other class of solution that I can think of is to spawn a thread of job that periodically sends a broadcast, but there are a couple of issues with this:
  1. Broadcasts work in this instance, but they aren't quite the right tool for the job. You have only one sender, one receiver, and you are both of them (It is a broadcast in name only).
  2. There isn't a good way to _stop_ these kinds of solutions when the component disconnects.

If we don't add this feature, I think users should avoid making `Thread` instances themselves and instead queue up ActiveJob jobs to periodically broadcast to them.

> Is it a good practice to handle the stopping of every timer you start, or will the timer stop automatically when the user loses the connection with the component?

Nope! Any timers created in the component will automatically be setup when it connects and cleaned up when it disconnects. This is already built-in to ActionCable.

My example has a contrived business logic requirement to stop the timer that I included specifically to demonstrate that it is possible.

> Are they able to stop a timer they start in a component somewhere else?

There is no public API for this, and I don't see a good place to add it since every component can have its own timer state.

> Is it odd that the user starts a periodic timer with every but turns it off with stop_periodic_timer? In rails, the two commands match more. I don't immediately have a better suggestion but can help come up with names if others agree.

The "real" (symmetric) API for this feature is `periodic_timer(name, handler = name, every:)` (for starting timers) and `stop_periodic_timer(name)` for removing them. `every(interval, handler, name: handler)` is just sugar for `periodic_timer`. I included it because I think it looks good, but if other people don't like it, we can take it out. I'm imagining that in most cases, uses will not manually call `stop_periodic_timer`.

**Comment by latortuga** [comment]
> There is no public API for this, and I don't see a good place to add it since every component can have its own timer state.

JavaScript has an API for timers already and we could mimic it. It would be a bit clumsy to communicate this to other components, you'd have to communicate using a global variable, the DB, or a broadcast of some kind. But at least the option would be available.

I'm thinking the API would use some kind of key related to the channel that the timer is communicated on.

```js
// JS timer
const id = setInterval(function() {
  alert("hi there")
}, 1000)

clearInterval(id)
```

**Comment by alecdotninja** [comment]
@latortuga The timers already have a name that can be used to refer to them inside of the component (it defaults to the method name of the handler).

```ruby
periodic_timer("timer_1", :handler, every: 5.minutes)
periodic_timer("timer_2", :handler, every: 5.minutes)
periodic_timer("timer_3", :handler, every: 5.minutes)

stop_periodic_timer("timer_1")
stop_periodic_timer("timer_2")
stop_periodic_timer("timer_3")
```

The only thing that is not supported is a global identifier, but I don't think this is a good idea. If you need to do something with a timer in response to something occurring in another part of the application, you can use broadcasts.

File: README.md [hunk 4]
```diff
@@ -112,25 +112,26 @@ class Todo < ApplicationModel
...
-    @todos = Todo.all.to_a
+  def initialize(count: 5)
+    @count = count
+    @todos = Todo.order(created_at: :desc).limit(count).pluck(:name)
   end
 
-  def handle_created
-    @todos = Todo.all.to_a
+  def handle_created(name)
+    @todos = [name, *@todos.first(@count - 1)]
```
Comment (latortuga) [inline README.md:134]: Perhaps this demo should be more like

```ruby
def handle_created(id)
  @todos = [Todo.find(id), *@todos.first(@count - 1)]
end
```

Comment (alecdotninja) [reply README.md:134]: I used the `name` because it avoids a database call for each component (which I wanted mostly for aesthetic reasons in the README).

**Pull request merged**